src/solver/multimarginloss/forward_reduced_multimarginloss.cpp

```diff
@@ -0,0 +1,237 @@
+            kernel.l_wk.push_back(xlocalsize);
+            kernel.l_wk.push_back(ylocalsize);
+            kernel.l_wk.push_back(zlocalsize);
+
+            kernel.g_wk.push_back(xgridsize);
+            kernel.g_wk.push_back(ygridsize);
+            kernel.g_wk.push_back(zgridsize);
+
+            result.construction_params.push_back(kernel);
+            _size = AlignUp(_size, LOCAL_SIZE_REDUCE) / LOCAL_SIZE_REDUCE;
+        } while(_size > 1);
+    }
+
+    result.invoker_factory = [](const std::vector<Kernel>& kernels) {
+        return [=](const Handle& handle_, const AnyInvokeParams& raw_params) {
+            decltype(auto) params = raw_params.CastTo<miopen::multimarginloss::InvokeParams>();
+            auto i_tv             = get_inner_expanded_tv<2>(deref(params.iDesc));
+            auto t_tv             = get_inner_expanded_tv<1>(deref(params.tDesc));
+            auto w_tv             = get_inner_expanded_tv<1>(deref(params.wDesc));
+
+            float elapsed = 0.0f;
+            HipEventPtr start;
+            HipEventPtr stop;
+
+            bool reset_profiling_state = false;
+            if(handle_.IsProfilingEnabled())
+            {
+                reset_profiling_state = true;
+                handle_.EnableProfiling(false);
+                start = miopen::make_hip_event();
+                stop  = miopen::make_hip_event();
+                hipEventRecord(start.get(), handle_.GetStream());
+            }
+            /* Phase 1: Calc loss for each element. */
+            {
+                decltype(auto) kernel = handle_.Run(kernels.front());
+                kernel(params.i,
+                       params.t,
+                       params.w,
+                       params.workspace,
+                       params.p,
+                       params.margin,
+                       params.divisor,
+                       i_tv,
+                       t_tv,
+                       w_tv);
+            }
+
+            /* Phase 2: Reduce */
+            auto size      = deref(params.iDesc).GetLengths()[0];
+            auto reduce_in = params.workspace;
+            auto reduce_out =
+                static_cast<Data_t>(static_cast<char*>(params.workspace) +
+                                    size * get_data_size(deref(params.oDesc).GetType()));
+
+            for(int i = 1; i < kernels.size(); ++i)
+            {
+                decltype(auto) kernel = handle_.Run(kernels[i]);
+                if(i + 1 != kernels.size())
+                {
+                    kernel(reduce_in, reduce_out, size);
+                    std::swap(reduce_in, reduce_out);
+                }
+                else
+                {
+                    kernel(reduce_in, params.o, size);
+                }
+                size = AlignUp(size, LOCAL_SIZE_REDUCE) / LOCAL_SIZE_REDUCE;
+            }
+
+            if(reset_profiling_state)
+                handle_.EnableProfiling(true);
+            if(handle_.IsProfilingEnabled())
+            {
```

Comment:
These two conditions can be squashed, see one of the previous comments in this file.

Reply:
changed like your recommendation